src/System.Net.Sockets/src/System/Net/Sockets/SafeSocketHandle.Unix.cs

```diff
@@ -358,6 +359,44 @@ internal static unsafe InnerSafeCloseSocket Accept(SafeSocketHandle socketHandle
                 res.SetHandle(acceptedFd);
                 return res;
             }
+
+            internal unsafe void TryUnblockSocket()
+            {
+                // Calling 'close' on a socket that has pending blocking calls (e.g. recv, send, accept, ...)
+                // may block indefinitly. This is a best effort attempt to not get blocked and make those operations return.
+                // We need to ensure we keep the expected TCP behavior that is observed by the socket peer (FIN vs RST close).
+                // What we do here isn't specified by POSIX, it works for Linux.
+
+                // Don't touch sockets which don't have the CLOEXEC flag set. These may be shared
+                // with other processes and we want to avoid disconnecting them.
+                int fdFlags = Interop.Sys.Fcntl.GetFD(this);
+                if (fdFlags == 0)
+                {
+                    return;
+                }
+
+                int type = 0;
+                int optLen = sizeof(int);
+                Interop.Error err = Interop.Sys.GetSockOpt(this, SocketOptionLevel.Socket, SocketOptionName.Type, (byte*)&type, &optLen);
+                Debug.Assert(err == Interop.Error.SUCCESS);
+                if (err == Interop.Error.SUCCESS)
+                {
+                    if (type == (int)SocketType.Stream)
+                    {
+                        // Connect to AF_UNSPEC causes an abortive close (TCP RST).
+                        Span<byte> address = stackalloc byte[32];
+                        SocketAddressPal.SetAddressFamily(address, AddressFamily.Unspecified);
+                        fixed (byte* pAddress = address)
+                        {
+                            Interop.Sys.Connect(this, pAddress, address.Length);
+                        }
+                    }
+                    else
+                    {
+                        Interop.Sys.Shutdown(this, SocketShutdown.Both);
```

Comment:
Would it make sense to call this regardless?  If Connect doesn't do what we expect, this would be a stop-gap, and if it does do what we expect, it would be a nop, right?

Reply:
It will cause a FIN close instead of a RST close on platforms that don't support connect AF_UNSPEC.
If we're ok with that, we can do it always.

Reply:
I'm not sure why RST would be important here. That is specific to connected TCP socket anyway, right? Let say process crashes and OS closes sockets: that would send FIN as well, right?

Reply:
> I'm not sure why RST would be important here. That is specific to connected TCP socket anyway, right?

This is matching the Windows behavior.

> Let say process crashes and OS closes sockets: that would send FIN as well, right?

Yes.